tfjs-backend-webgpu/src/kernels/from_pixels_webgpu.ts

```diff
@@ -26,35 +26,123 @@ export class FromPixelsProgram implements WebGPUProgram {
   variableNames = ['A'];
   dispatchLayout: {x: number[]};
   dispatch: [number, number, number];
+  pipeline: GPUComputePipeline;
+  bindGroupLayout: GPUBindGroupLayout;
+
+  uniform: GPUBuffer;
+  lastUniformContent = [0, 0, 0];
+
+  inputTexture: GPUTexture = null;
+  lastPixelSize = {width: 0, height: 0};
 
   constructor(outputShape: number[]) {
     this.outputShape = outputShape;
     this.dispatchLayout = flatDispatchLayout(this.outputShape);
     this.dispatch = computeDispatch(this.dispatchLayout, this.outputShape);
-
+    const workGroupSize = 256;
     this.userCode = `
-      void main() {
-        ivec3 coords = getOutputCoords();
-        int texR = coords[0];
-        int texC = coords[1];
-        int depth = coords[2];
-        vec2 uv = (vec2(texC, texR) + halfCR) / vec2(outShape.yx);
-
-        vec4 values = texelFetch(A, uv);
-        float value;
-        if (depth == 0) {
-          value = values.r;
-        } else if (depth == 1) {
-          value = values.g;
-        } else if (depth == 2) {
-          value = values.b;
-        } else if (depth == 3) {
-          value = values.a;
-        }
+    layout (local_size_x = ${workGroupSize},
+      local_size_y = 1,
+      local_size_z = 1) in;
+    layout(set = 0, binding = 1, rgba8) uniform readonly image2D srcImage;
+    layout(set = 0, binding = 2) uniform Meta {
+      int size;
+      int width;
+      int numChannels;} outShape;
 
-        setOutput(floor(value * 255.0 + 0.5));
+    void main() {
+      ivec3 coords = getCoordsFromFlatIndex(
+        int(gl_GlobalInvocationID.x) * outShape.numChannels);
+      int texR = coords[0];
+      int texC = coords[1];
+      int depth = coords[2];
+      vec4 values = imageLoad(srcImage, ivec2(texC, texR));
+      float value;
+      for(int i = 0; i < outShape.numChannels; i++) {
+        value = values[i];
+        int flatIndex =
+          int(gl_GlobalInvocationID.x) * outShape.numChannels + i;
+        if (flatIndex < outShape.size) {
+          result[flatIndex] = int(floor(255.0 * value));
+        }
       }
+
+    }
     `;
     this.shaderKey = 'fromPixel';
   }
+
+  setWebGPUBinary(
+      bindGroupLayout: GPUBindGroupLayout, pipeline: GPUComputePipeline) {
+    this.bindGroupLayout = bindGroupLayout;
+    this.pipeline = pipeline;
+  }
+
+  setUniform(device: GPUDevice, uniformData: number[]) {
+    const [uniformBuffer, uniformMapping] =
+        device.createBufferMapped({size: 4 * 3, usage: GPUBufferUsage.UNIFORM});
+    new Uint32Array(uniformMapping).set(uniformData);
+    uniformBuffer.unmap();
+    this.uniform = uniformBuffer;
+  }
+
+  getInputTexture(device: GPUDevice, pixelWidth: number, pixelHeight: number):
```

Comment:
there is a webgpu buffer manager (buffer_manager.ts) that takes care of recycling buffers whenever possible and allocating new buffers - maybe we could introduce a similar mechanism for textures

Reply:
Yes, agree. But for fromPixel, the inputTexture won't be created/recycle in high frequency. (And I think other ops don't use texture) So I think maybe for fromPixels, we can keep this simple management WDYT?